javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll

```diff
@@ -0,0 +1,113 @@
+/**
+ * For internal use only.
+ *
+ * Defines a set of characteristics that a particular endpoint might have. This set of characteristics is used to make
+ * decisions about whether to include the endpoint in the training set and with what label, as well as whether to score
+ * the endpoint at inference time.
+ */
+
+import experimental.adaptivethreatmodeling.EndpointTypes
+import semmle.javascript.security.dataflow.SqlInjectionCustomizations
+private import semmle.javascript.security.dataflow.DomBasedXssCustomizations
+private import semmle.javascript.security.dataflow.NosqlInjectionCustomizations
+private import semmle.javascript.security.dataflow.TaintedPathCustomizations
+
+abstract class EndpointCharacteristic extends string {
+  // The name of the characteristic, which should describe some characteristic of the endpoint that is meaningful for
+  // determining whether it's a sink and if so of which type
+  bindingset[this]
+  EndpointCharacteristic() { any() }
+
+  // Indicators with confidence at or above this threshold are considered to be high-confidence indicators.
+  float getHighConfidenceThreshold() { result = 0.8 }
```

Comment:
One question I had actually was what is the role of this predicate? I'm guessing it's a categorical selection wrapper on the more fine grain confidence floats, but why do we need it at this stage? (same question with the `Medium` one below)

Reply:
You're right, I could have left these for a later PR. I use them in the endpoint selection code. For example, to implement logic such as "if the list of characteristics includes positive indicators with high confidence for this class, select this as a training sample belonging to the class". I put them in `EndpointCharacteristic` because I think the place that sets confidences for various types of endpoints should also define what we mean by "high confidence".

Reply:
Should we introduce them in the PR that will use them then? I know you have written more code locally, but for us that will make things a little easier to follow along.

Reply:
IMHO it's not worth deleting them from this PR just to add them in the next PR 🤷

Reply:
ok but in the future please do not introduce concepts not discussed beforehand and not used in the PR you open.